### PR TITLE
port changes from lightning

### PR DESCRIPTION
**Pull request by timmysilv**
Changes ported over from PennyLaneAI/pennylane-lightning#424

Quite a bit had to be re-jigged a bit, but the test I added runs `dev.adjoint_jacobian` on various tapes with old-school and new-school observables and ensures that they are the same, which seems promising to me!

Note: I linked the GPU PR to this one below, but I'll address reviews here and port over whatever's needed before requesting a review on that one.

### CODE REVIEW

File: pennylane_lightning_kokkos/_serialize.py [hunk 1]
```diff
@@ -209,12 +200,16 @@ def _serialize_ob(ob, wires_map, use_csingle):
...
+    elif ob._pauli_rep is not None:
+        return _serialize_pauli_sentence(ob._pauli_rep, wires_map, use_csingle)
     elif ob.name == "Hermitian":
         raise TypeError(
-            f"Hermitian observables are not currently supported for adjoint differentiation. Please use Pauli-words only."
+            "Hermitian observables are not currently supported for adjoint differentiation. Please use Pauli-words only."
         )
     else:
-        return _serialize_named_ob(ob, wires_map, use_csingle)
+        raise TypeError(f"Unknown observable found: {ob}. Please use Pauli-words only.")
```
Comment (timmysilv) [inline pennylane_lightning_kokkos/_serialize.py:212]: the key difference from the lightning change ([those lines](https://github.com/PennyLaneAI/pennylane-lightning/blob/769997cdce4096b933129d421fc703dc9d728f37/pennylane_lightning/_serialize.py#L157-L158), for reference) is in these lines.

In pennylane-lightning, if an observable isn't a Hamiltonian, Tensor, Named, or something with a pauli rep, it just uses `Hermitian` which calls `qml.matrix`.

Here, if it doesn't meet any of the other requirements, it just raises an error. Lightning uses `_serialize_ob` even for regular expval, but this repo (and GPU) only serializes obs for adjoint differentiation, which I guess the C++ hermitian class doesn't support?

I gave a new error if the observable isn't support and it's not a `qml.Hermitian`, but we could combine them if we want.

**Review by AmintorDusko — APPROVED** [review]
Changes here are very similar to what was reviewed for Lightning.
Thank you for that.

File: pennylane_lightning_kokkos/_serialize.py (hunk 1 above)
Comment (mlxd) [reply pennylane_lightning_kokkos/_serialize.py:212]: I'll take a quick look at this --- Hermitian should be supported here in the adjoint execution pipeline, though only for non-parametric gate evaluation.

Comment (mlxd) [reply pennylane_lightning_kokkos/_serialize.py:212]: Just checked, this shouldn't be disabled. I'll make a PR that fixes this --- we can then replicate the same dispatch function as with lightning.qubit and remove the error.

**Review by vincentmr — APPROVED** [review]
Thanks @timmysilv , looks good to me.

File: tests/test_adjoint_jacobian.py [hunk 2]
```diff
@@ -22,7 +22,13 @@
 from pennylane import QNode, qnode
 from scipy.stats import unitary_group
 import pennylane_lightning_kokkos as plk
-from pennylane_lightning_kokkos.lightning_kokkos_qubit_ops import InitArguments
+from pennylane_lightning_kokkos.lightning_kokkos_qubit_ops import (
+    InitArguments,
+    NamedObsKokkos_C64,
```
Comment (mlxd) [inline tests/test_adjoint_jacobian.py:27]: Can these be C128 (as in std::complex) as well, since that is the default we expect throughout.

**Review by mlxd — APPROVED** [review]
Thanks for these additions @timmysilv 
Happy to merge it in.

**Pull request merged**